File: 0797-All-Paths-From-Source-to-Target/0797-All-Paths-From-Source-to-Target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        all_paths = []
        def dfs(node, path):
            if node == n - 1:
                all_paths.append(path)
                return
            for neighbor in graph[node]:
                dfs(neighbor, path + [neighbor])
        dfs(0, [0])
        return all_paths
#TC: O(V + E), where V is the number of vertices (nodes) in the graph and E is the number of edges.
#SC: O(V + E), since the list of all paths and the current path passed to the recursive function both have length at most V + E.

File: 0797-All-Paths-From-Source-to-Target/0797-All-Paths-From-Source-to-Target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<vector<int>> all_paths;
        function<void(int, vector<int>)> dfs;
        dfs = [&](int node, vector<int> path) {
            if (node == n - 1) {
                all_paths.push_back(path);
                return;
            }
            for (int neighbor : graph[node]) {
                vector<int> new_path = path;
                new_path.push_back(neighbor);
                dfs(neighbor, new_path);
            }
        };
        dfs(0, {0});
        return all_paths;   
    }
};
